voice.go

```
@@ -231,314 +201,508 @@ type VoiceSpeakingUpdate struct {
 		}
 	}
+
+	v.failure(ErrVoiceReconnectionLimit)
 }
 
 // wsEvent handles any voice websocket events. This is only called by the
 // wsListen() function.
-func (v *VoiceConnection) onEvent(message []byte) {
+func (v *VoiceConnection) onEvent(ctx context.Context, binary bool, message []byte) {
 
-	v.log(LogDebug, "received: %s", string(message))
-
-	var e Event
-	if err := json.Unmarshal(message, &e); err != nil {
-		v.log(LogError, "unmarshall error, %s", err)
-		return
+	if binary {
+		v.log(LogDebug, "received binary: %x", message)
+	} else {
+		v.log(LogDebug, "received string: %s", string(message))
 	}
 
-	switch e.Operation {
+	if binary {
+		// TODO: implement DAVE
+	} else {
 
-	case 2: // READY
-
-		if err := json.Unmarshal(e.RawData, &v.op2); err != nil {
-			v.log(LogError, "OP2 unmarshall error, %s, %s", err, string(e.RawData))
+		var e voiceWebsocketMessage
+		if err := json.Unmarshal(message, &e); err != nil {
+			v.log(LogError, "unmarshall error, %s", err)
 			return
 		}
 
-		// Start the voice websocket heartbeat to keep the connection alive
-		go v.wsHeartbeat(v.wsConn, v.close, v.op2.HeartbeatInterval)
-		// TODO monitor a chan/bool to verify this was successful
+		if e.Sequence != nil {
 
-		// Start the UDP connection
-		err := v.udpOpen()
-		if err != nil {
-			v.log(LogError, "error opening udp connection, %s", err)
-			return
+			v.Cond.L.Lock()
+			v.seqAck = *e.Sequence
+			v.Cond.L.Unlock()
 		}
 
-		// Start the opusSender.
-		// TODO: Should we allow 48000/960 values to be user defined?
-		if v.OpusSend == nil {
-			v.OpusSend = make(chan []byte, 2)
-		}
-		go v.opusSender(v.udpConn, v.close, v.OpusSend, 48000, 960)
+		switch e.Operation {
 
-		// Start the opusReceiver
-		if !v.deaf {
-			if v.OpusRecv == nil {
-				v.OpusRecv = make(chan *Packet, 2)
+		case 2: // READY
+			op2 := voiceOP2{}
+
+			if err := json.Unmarshal(e.RawData, &op2); err != nil {
+				err := fmt.Errorf("OP2 unmarshal error, %s, %s", err, string(e.RawData))
```

it's recommended to use `%w` with [fmt.Errorf](https://pkg.go.dev/fmt#Errorf) when you return it to client/user so they can match it with `errors.Is` and act on it. Otherwise you are hiding the underlying error type
```suggestion
				err := fmt.Errorf("OP2 unmarshal error, %w, %s", err, string(e.RawData))
```